### ISSUE TITLE
New feature: annotate all factory methods with @CheckReturnValue

### ISSUE THREAD

**Issue by TWiStErRob** [android/android-test#1325]
### Description
To prevent situations like: https://twitter.com/yogurtearl/status/1347344454668062722

We could leverage a best effort annotation that tools could detect.
Using `androidx.annotation.CheckResult` would be ideal, because (hopefully) lint would pick up on that.
If lint doesn't [Errorprone can](https://errorprone.info/bugpattern/CheckReturnValue), so an alternative is JSR305's `javax.annotation.CheckReturnValue`. Best is both, actually this already exists on Espresso.onView.

### Steps to Reproduce
In a sudden lapse of judgement, import the wrong thing:
```java
import static androidx.test.espresso.action.ViewActions.closeSoftKeyboard;

@Test public void test() {
    ...
    closeSoftKeyboard();
    ...
}
```
of course the right one is:
```java
import static androidx.test.espresso.Espresso.closeSoftKeyboard;
```

### Expected Results
Some kind of warning when using Espresso wrong.

### Actual Results
Silent nothing, many hours of dev time wasted.

### AndroidX Test and Android OS Versions
N/A

### Link to a public git repo demonstrating the problem:
N/A

### Proposal

Annotate everything in:
 * *View*Actions.*
 * androidx.test.espresso.matcher.*
 * androidx.test.espresso.intent.Intents.intending
 * potentially others

I would be happy to contribute the adding of annotations on the Espresso public surface, but I have a feeling internally in Google you might require some cleanup in the monorepo if this ever happened (i.e. remove side-effect-less calls, if any). Although those can be discovered on this branch, and committed before merge.

**Comment by TWiStErRob** [comment]
It's a start :) https://github.com/android/android-test/pull/1334

**Comment by TWiStErRob** [comment]
@brettchabot I would really go for the androidx annotation, because errorprone is a third party we need to explicitly install, lint reaches more people, out of the box.

**Comment by TWiStErRob** [comment]
cc @cpovirk

**Comment by cpovirk** [comment]
Thanks for CCing me!

My understanding is that Android Lint recognizes the Error Prone `@CheckReturnValue`, too. (Hmm, but maybe Error Prone does _not_ recognize the `androidx` `CheckResult`? I should look into that....) So I think we can use that one, or maybe you're aware of other considerations?

I'm putting together another change for `ViewActions` now. It does indeed turn up some bugs in our repo. Thanks for the list!

**Comment by cpovirk** [comment]
> Hmm, but maybe Error Prone does _not_ recognize the `androidx` `CheckResult`? I should look into that....

It does not. There is feature request for this in our internal tracker... from 6 years ago... :)

The good news is that the reason I'm here is that some other people on my team are putting a bunch of work into `CheckReturnValue`/`CheckResult`-type checks, so one of them might well be interested in picking it up.

**Comment by cpovirk** [comment]
Whoa, `Intents.intending` is probably the second-most misused API I've seen yet in this project :)

**Comment by TWiStErRob** [comment]
> My understanding is that Android Lint recognizes the Error Prone @CheckReturnValue, too. 

I just double-checked [`CheckResultDetector`](https://android.googlesource.com/platform/tools/base/+/studio-master-dev/lint/libs/lint-checks/src/main/java/com/android/tools/lint/checks/CheckResultDetector.kt#49) in lint and it has supported 7 aliases in AGP 3.2.1-7.1.3 for sure.

> So I think we can use that one, or maybe you're aware of other considerations?

Sounds like simply adding the error prone annotation would suffice then.
If lint hadn't supported it, there would be considerations, but I think this is fine. Using it would reach a very wide audience already using lint and/or errorprone as soon as they update.

I can't think of any other issues, maybe @jakewharton, @jrodbx has ideas for what needs to be considered (as library authors and community champions).

> Thanks for the list!

No probs, but please review all the public methods, because I only spent the one minute writing them out here.

**Comment by JakeWharton** [comment]
These libraries already have a dependency on the AndroidX annotations artifact whose `@CheckResult` annotation is already honored by default in the tooling that ships with the standard build system for Android projects. I do not think it's a good idea to add error prone's annotations solely for the purpose of adding an annotation that's already present in a dependency simply because Google's custom static analysis tool originally built for their non-Android monorepo doesn't yet honor it.

So TL;DR my vote is: use the Android standard annotation and fix error-prone separately.

**Comment by cpovirk** [comment]
This is all sadly reminiscent of our recent nullness work, in which the most standard annotation is not necessarily the one that produces the best experience for users :(

This library already had the Error Prone annotations in its deps, too, so that may eliminate the main downside for users to using that. (That's what https://github.com/android/android-test/pull/1334 used, and so will the `ViewActions` and `Intentions.intending` changes I have in flight.) If there are other downsides, though, let me know. And I'm not on this particular project; we're just poking our heads into random Google codebases, as per usual :)

**Comment by cpovirk** [comment]
Sorry, I didn't conclude that last thought. I was just going to say that the actual owners may have other reasons to prefer one or the other that I wouldn't know about.

**Comment by JakeWharton** [comment]
If both dependencies are already present then yes I don't think it matters at all. 

Hot take: all non-void functions should have required checking the result as enforced by the language compiler by default with an opt-out rather than an opt-in!

**Comment by TWiStErRob** [comment]
Is it possible errorprone was added by one of these random pokes in the past? There are only [3 usages](https://github.com/android/android-test/search?q=com.google.errorprone.annotations), you added the 4th, @cpovirk.

Compared to 250 usages of [androidx.annotation](https://github.com/android/android-test/search?q=androidx.annotation&type=code)

I like your hot-take, Jake, and agree; sadly #wrongforum, is there a KT ticket for that I can vote on?

**Comment by cpovirk** [comment]
> Is it possible errorprone was added by one of these random pokes in the past?

Very likely. I don't know if the _other_ Error Prone annotations have equivalents in other artifacts. Maybe they do?

(I do see a few instances of `@CheckReturnValue @CheckResult`... :))

RE: Kotlin:

I don't know if there's a Kotlin ticket for changing the language default, but there's [KT-12719](https://youtrack.jetbrains.com/issue/KT-12719) for building opt-in enforcement into the language.

(We are in fact trying to flip the default inside Google. It's turning up a lot of fun bugs.)

**Comment by TWiStErRob** [comment]
Well I see 4 usages:
 * com.google.errorprone.annotations.MustBeClosed in Tracer
   A library-internal class, so consumers don't really use it.
 * com.google.errorprone.annotations.MustBeClosed in Tracing
   An @ExperimentalTestApi public method,
   which is tightly coupled with the library-internal Tracer interface, so no end-user can use it.
   Only useful if the consumer has errorprone... (I'm yet to actively work on a project that uses it.)
 * com.google.errorprone.annotations.InlineMe in InstrumentationRegister
   A deprecated public class, useless without tooling support, I'd be surprised in IDEA supported this.
   Kotlin has an equivalent `@InlineMe(replacement = "...") @Deprecated` -> `@Deprecated(replaceWith = ReplaceWith("..."))`, which could be used if the deprecated class was Konverted which looks feasible in a binary-compatible way.
 * com.google.errorprone.annotations.CheckReturnValue in IntentMatchers, the one you just added

So based on this, it's removable with almost no effort/drawback.

**Comment by cpovirk** [comment]
Removing the direct dep is likely doable. We could even set things up to continue to use those annotations inside our own repo if we wanted.

Still, I'm not sure that this is going to make much difference to users: I think that the Error Prone annotations get pulled it transitively by a number of other deps -- Dagger, Flogger, Guava, Guice, Truth:

https://github.com/android/android-test/blob/41855232b54fb77af11744e446d11204749b789c/WORKSPACE#L108-L120

The dream is that someday _none_ will need that dep, but I would expect it to be many years before that happens.

**Comment by cpovirk** [comment]
My mistake, you're right about Dagger: https://repo1.maven.org/maven2/com/google/dagger/dagger/2.41/dagger-2.41.pom

**Comment by cpovirk** [comment]
(And yes, deps of the few androidx-test artifacts that I finally got around to stop-checking do look fairly minimal (e.g., [core](https://maven.google.com/androidx/test/core/1.4.1-alpha05/core-1.4.1-alpha05.pom), [junit](https://maven.google.com/androidx/test/ext/junit/1.1.4-alpha05/junit-1.1.4-alpha05.pom)).)

**Comment by sgjesse** [comment]
On the use of `com.google.errorprone.annotations.InlineMe` in `InstrumentationRegistry`, I have just filed https://issuetracker.google.com/260089912, where R8 will fail on missing classes when processing this library.
